haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,216 @@
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    |
+    |   # Users can also provide a list of questions/document dictionary pairs to generate pseudo labels for
+    |
+    |   question_doc_pairs: List[Dict[str, str]] = [{"question": "question text ...", "document": "document text ..."}]
+    |   plg = PseudoLabelGenerator(question_doc_pairs, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    ```
+    """
+
+    def __init__(
+        self,
+        question_producer: Union[QuestionGenerator, List[Dict[str, str]]],
+        retriever: BaseRetriever,
+        cross_encoder_model_name_or_path: str = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        total_number_of_questions: int = sys.maxsize,
+        top_k: int = 10,
+    ):
+        """
+        Loads the cross encoder model and prepares PseudoLabelGenerator.
+
+        :param question_producer: The question producer used to generate questions or a list of already produced
+        questions/document pairs in Dict format {"question": "question text ...", "document": "document text ..."}.
+        :type question_producer: Union[QuestionGenerator, List[Dict[str, str]]]
+        :param retriever: The retriever used to query document stores
+        :type retriever: BaseRetriever
+        :param cross_encoder_model_name_or_path: The path to the cross encoder model, defaults to
+        cross-encoder/ms-marco-MiniLM-L-6-v2
+        :type cross_encoder_model_name_or_path: str (optional)
+        :param total_number_of_questions: This is the total number of questions that will be generated, defaults
+        to sys.maxsize
+        :type total_number_of_questions: int
+        :param top_k: This is the number of answers that you want to retrieve for each question, defaults to 10
+        :type top_k: int (optional)
+        """
+
+        super().__init__()
+        self.question_document_pairs = None
+        self.question_generator = None
+        if isinstance(question_producer, QuestionGenerator):
+            self.question_generator = question_producer
+        elif isinstance(question_producer, list) and len(question_producer) > 0:
+            example = question_producer[0]
+            if isinstance(example, dict) and "question" in example and "document" in example:
+                self.question_document_pairs = question_producer
+            else:
+                raise ValueError("question_producer list must contain dicts with keys 'question' and 'document'")
+        else:
+            raise ValueError("Provide either a QuestionGenerator or nonempty list of questions/document pairs")
+
+        self.retriever = retriever
+        self.cross_encoder = CrossEncoder(cross_encoder_model_name_or_path)
```

Comment:
I think you will have to pass in `default_activation_function=nn.Identity` when initializing the CrossEncoder. This is not described in the paper, but for some tables and pictures of the labels and margins that were used you can guess that they used raw CrossEncoder scores and not the ones that were passed through an `nn.Sigmoid` activation function.

Reply:
Again, here I followed Niels' notebook. LMK what you think about it.

Reply:
Hm, I don't know, in the paper they write: `Here the dot-product is usually used due to the infinite range of the cross-encoder score` -> default behaviour of `CrossEncoder` is to pass it through nn.Sigmoid which does not produce an infinite range.

Reply:
Ok, very good point @mathislucka . I'll double-check with Kexin.

Reply:
@mathislucka and I verified that by default `default_activation_function` gets set to nn.Identity when initializing the CrossEncoder